examples/medplum-demo-bots/src/deduplication/find-matching-patients.ts

```diff
@@ -0,0 +1,118 @@
+import { BotEvent, MedplumClient, createReference, getReferenceString, resolveId } from '@medplum/core';
+import { ListEntry, Patient, RiskAssessment, Task } from '@medplum/fhirtypes';
+
+/**
+ * This Bot listens for changes to a `Patient` resource and searches for potential patient duplicates.
+ * The matching algorithm used is to search for exact matches:
+ *  - first name
+ *  - last name
+ *  - date of birth
+ *  - postal code
+ *
+ * See: https://www.medplum.com/docs/fhir-datastore/patient-deduplication#matching-rules
+ * for more potential matching rules
+ *
+ * @param medplum - The Medplum client instance.
+ * @param event - The BotEvent containing the Patient resource.
+ */
+export async function handler(medplum: MedplumClient, event: BotEvent<Patient>): Promise<void> {
+  //This bot should only be triggered by a Patient resource Subscription only
+  const srcPatient = event.input;
+
+  // Read the patient's "Do Not Match" list. Assume there is only one applicable list
+  const doNotMatchList = await medplum.searchOne('List', {
+    code: 'http://example.org/listType|doNotMatch',
+    subject: getReferenceString(srcPatient),
+  });
+
+  console.info('Searching for matches for: ' + getReferenceString(srcPatient));
+
+  // Search for potential active duplicate patients by matching first name, last name, birthdate, and postal code.
+  const candidateMatches = await medplum.searchResources('Patient', {
+    name: srcPatient.name?.[0].family,
```

Comment:
Do these also need safe chaining operators through the right side of the expression?  E.g. `srcPatient.name?.[0]?.family`